Fixes #32224 - Expose Katello events for Webhooks

ofedoren opened the pull request:
This PR exposes new events for subscriprions within Foreman Webhooks plugin, to test this you might want to have it installed.

@lzap, you might want to test it and probably tell what is missing.

TODO:
 - [x] Add DSL docs
 
 NOTE: I've changed the order of initializers because currently the actions are being loaded after all plugin registration methods are called, which leads to addition of new observable events being impossible. Also `extend_observable_events` is available from `2.4`, so I've updated the required version as well.

lzap commented:
@jjeffers what are our chances getting this into 4.0? cheers :-)

jjeffers commented:
> @jjeffers what are our chances getting this into 4.0? cheers :-)
> 
> I will be testing this tomorrow.

Probably low chance, since we just built rc3.1. The next release should be the GA.

lzap commented:
Yeah that's what I thought. I was also hoping the patch would be even smaller, it looks like an initializer change is needed. Okay.

lzap commented on lib/katello/engine.rb:
```diff
@@ -99,6 +89,17 @@ class Engine < ::Rails::Engine
+      require 'katello/permissions'
     end
 
     initializer "katello.set_dynflow_middlewares", :before => :finisher_hook do |_app|
```
I am not sure I understand the consequences of this, but since I'd like to ask for minor release inclusion for this patch I wonder if @adamruzicka would be able to confirm this is safe to do.

adamruzicka replied:
This looks reasonable to me and I don't see any immediate footguns. @jlsherrill anything I'm overlooking?

lzap commented on app/lib/actions/katello/actions_jail.rb:
```diff
@@ -0,0 +1,33 @@
+      end
+    end
+  end
+end
```
I am wondering how a concern will work in this case, class Jail is defined via the concern and then redefined later. I guess I need to test it.

By the way is the reason you inherit `::Actions::ObservableAction::Jail` instead of `Safemode::Jail` to also have `task` allowed?

ofedoren replied:
Well, I guess for actions we want have `task` object available to get some info. The first action that has that is `host_run_job_succeeded`, so I thought it might be useful for all the actions, thus I inherit `::Actions::ObservableAction::Jail`.

Also, I've updated the way the `Jail` gets possibility to be re-opened after the module is included.

lzap commented:
Nicely done, can you add those DSL docs? I am starting with testing right away. I was hoping that we only add a single line into plugin.rb that initializer reordering is unfortunate but I can test some edge-case scenarios (e.g. dynflow resume?).

ofedoren commented:
@lzap, working on the docs, I've updated the PR so it's testable with Safemode enabled (I forgot that when I was writing this I had that disabled)

lzap commented on app/lib/actions/katello/actions_jail.rb:
```diff
@@ -0,0 +1,40 @@
+          end
+          base.const_set('Jail', jail_class)
+        end
+      end
```
I am just wondering if it is not cleaner just to define attributes in this module and simply repeating all allowed attributes for each class. But I am okay with this.

ofedoren replied:
Sure it would be cleaner than some magic Ruby-woodoo, but repeating is not so good as well since one might forget to update allowed methods somewhere else.

lzap commented:
Rubocop failures.

ofedoren commented:
@lzap, updated. Test failures don't seem to be related..

lzap commented on app/lib/actions/katello/actions_common_jail.rb:
```diff
@@ -0,0 +1,62 @@
+      end
+    end
+  end
+end
```
This really calls for breaking up into two concerns: `ActionsContentViewJailConcern` and `ActionsOrganizationJailConcern`.

lzap reviewed (changes requested):
Other than that, I am good.

ofedoren commented:
@lzap, updated. I've renamed those a bit to have a structure.

Just a reminder for reviewers: if you want to test docs locally as well:

`FOREMAN_APIPIE_LANGS=en bundle exec rake apipie_dsl:cache` within `foreman` directory.

lzap reviewed (approved):
Thank you I am good.

To Katello reviewers: This patch introduces Jail support for Safemode, in order to expose methods they must be explicitly listed. We tried to find minimum amount of methods which gives reasonable power to webhooks users. We have landed on this rather flat list of methods because the dynflow actions carries the data in a hash.

lzap commented on app/lib/actions/katello/jail_concern/content_view.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+    end
+  end
+end
```
I think I found a bug tho, I am getting safemode error for Content View create event, I think this concern is not included in the ContentView model at all. It is not a Dynflow action but ActiveRecord event as well, so it should be just a simple Jail.

```
2021-04-06T12:24:34 [I|app|06a8fc9e] ForemanWebhooks::EventSubscriber: content_view_created.event.foreman event received
2021-04-06T12:24:34 [E|app|06a8fc9e] Safemode::NoMethodError: undefined method '#content_view_id' for Katello::ContentView::Jail (Katello::ContentView)
```

ofedoren replied:
Well, for content view object it should be just `id`, not the `content_view_id`. There is (or at least should be, I don't remember) documentation for Content View class in `all`/`additional` section in DSL docs.

lzap replied:
So it is used in other Dynflow action(s) but in Content View (CRUD) the Jail is missing, so actually here it is correct.

lzap commented on app/models/katello/content_view.rb:
```diff
@@ -86,6 +87,8 @@ class ContentView < Katello::Model
+
     def self.in_environment(env)
       joins(:content_view_environments).
         where("#{Katello::ContentViewEnvironment.table_name}.environment_id = ?", env.id)
```
I think organization and location (name,label,id) is worth exposing too for CV.

lzap replied:
We are missing Jail for this model here.

lzap reviewed (changes requested):
Actually one missing jail.

ofedoren commented:
@lzap, updated. There was Jail for content views already, I've just updated it with `id` and `organization`.

lzap commented:
> @lzap, updated. There was Jail for content views already, I've just updated it with `id` and `organization`.

Oh my bad I started testing with `id` and I assumed there was no Jail class. Name would be maybe enough, but this is good, let me retest.

lzap commented on app/lib/actions/katello/content_view/promote.rb:
```diff
@@ -2,6 +2,11 @@ module Actions
+        extend ApipieDSL::Class
+        include ::Actions::ObservableAction
+        include Actions::Katello::JailConcern::Organization
+        include Actions::Katello::JailConcern::ContentView
```
Just for the record, this does not work atm, organization jail is overwritten in the include block by content view.

lzap replied:
Since the Jail.allow method concatenates to a list, you can call it multiple times. So I suggest you assume Jail already exists, move these includes down after Jail class definition and append from the included block.

ofedoren replied:
I've tried that one, but when I try something like `base::Jail` or `base.const_get(:Jail)` it returns an object that doesn't understand `.inspect` even.

lzap reviewed (changes requested):
We are close, it does not work yet.

ofedoren commented:
@lzap, thanks for noticing that `const_set` will replace the `Jail` with the second `include`. I've played a little bit more with class reopening and noticed that what works with plain Ruby in `irb` doesn't work in Rails. Well, even if it worked I agree that this complicates simple things and might break something in the future.

I've returned the simple way for `Jail`s and left mixins only for their main purpose.

lzap reviewed (approved):
Everything works as expected. Can we have somebody looking on this PR, there are many comments already so folks might be skipping it. @ehelms @jlsherrill thanks.

ehelms commented on lib/katello/engine.rb:
```diff
@@ -99,6 +89,17 @@ class Engine < ::Rails::Engine
+      require 'katello/plugin'
+
+      # extend builtin permissions from core with new actions
+      require 'katello/permissions'
```
I am confused by the moving around of this code. It appears to not change anything other than where it appears in the file?

ofedoren replied:
Yes. From that perspective there are no changes at all. But the order does affects if the needed constants are available in plugin registration block.

ehelms commented on app/lib/actions/katello/repository/sync.rb:
```diff
@@ -124,6 +126,62 @@ def regenerate_applicability(repo, contents_changed)
+        def sync_result
+          input['sync_result']
+        end
+        apipie :class, "A class representing #{self} object" do
```
Is this tied to `def sync_result` ? Otherwise, a blank line above this would help.

ofedoren replied:
Noticed, I've added missing blank lines.

ehelms commented on app/lib/actions/katello/repository/sync.rb:
```diff
@@ -124,6 +126,62 @@ def regenerate_applicability(repo, contents_changed)
+                :repository_id, :repository_name, :repository_label,
+                :product_id, :product_name, :product_label,
+                :sync_result, :contents_changed
+        end
```
Do I understand right that the `apipie` definition, the include JailConcern and the class Jail declaration are all tied to the same concept? Events?

ofedoren replied:
Yes, for `actions` these are needed to fully support events in Foreman Webhooks that are being fired up if e.g. repo sync has happened.

`apipie` block is for documentation; JailConcern with public methods to use in webhook templates; Jails are to make those methods available in safe mode.

ehelms replied:
Thanks. It seems like it would be useful to group all of the pieces related to webhooks together somehow but I'll let Katello core maintainers weigh in on the overall code structure. I appreciate the helpful details understanding. I was worried that it would be hard over time to track what is required to ensure Webhooks work (and subsequently do not break in the future) and what aspects are unrelated. Also, how to help developers who want to add additinal webhook "hooks" into the code know what to add.

ehelms commented:
> To Katello reviewers: This patch introduces Jail support for Safemode, in order to expose methods they must be explicitly listed. We tried to find minimum amount of methods which gives reasonable power to webhooks users. We have landed on this rather flat list of methods because the dynflow actions carries the data in a hash.

Is the addition of Jail support for Safemode directly tied to Webhooks support? Is that something that is a pre-requisite? Does Jail support without "events" make sense for Katello? One thing I am trying to understand from this while looking over the code is as a developer what parts of the changes are what I need to add to produce events? (and which parts are not 'directly' required)

ofedoren commented:
@ehelms, short answer would be:

If a developer wants to expose events to be subscribed after then simple `include Foreman::ObservableModel` for models or `include ::Actions::ObservableAction` for actions should be enough. Other additions are related, but not required. Although without them there is no much of usage from the user end perspective since everything will be internal.

lzap commented:
I can confirm what @ofedoren said, making events observable via foreman_webhooks is really that simple and we have documented this in the devel docs: https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc#adding-subscribers

However if you turn on safemode, then the data coming in the webhook templates will be useless because there will be no way of getting any data out of it. Therefore the Jail and apidocs portions of the patch.

jlsherrill commented:
Wanted to update, i'm in the middle of testing this.  Overall looks good, will update again soon.

jlsherrill reviewed (approved):
ACK, tested the events and they all worked, rendering the template i created in each case.  I'm okay with the overall structure of the code here

lzap commented:
Thanks, I asked for 4.0.1 in the issue.

lzap commented:
I see Katello 4.0.1 being pushed to rubygems. I did set target release in RM to at least have a conversation about possibility of this patch to be cherry picked. Now I've sent my presentation recording to Foreman Birthday party where I announce Foreman Webhooks with Katello 4.0 support :-(

https://projects.theforeman.org/versions/1369

Should I move the target release to 4.0.2 with hope it would make the cut? :-)

ofedoren commented:
@lzap isn't it already there? https://github.com/Katello/katello/commits/KATELLO-4.0 / https://github.com/Katello/katello/compare/4.0.0...4.0.1

jjeffers commented:
> @lzap isn't it already there? https://github.com/Katello/katello/commits/KATELLO-4.0 / [4.0.0...4.0.1](https://github.com/Katello/katello/compare/4.0.0...4.0.1)

It should be in there - please see https://github.com/Katello/katello/pull/9359

lzap commented:
Holy smokes, I think I've used all of them. THANKS! :-)
